Factor out PSBT utilities from RPCs for use in GUI code; related refactoring.

gwillen opened the pull request:
* Move most PSBT definitions into psbt.h.
* Move most PSBT RPC utilities into psbt.{h,cpp}.
* Move wallet-touching PSBT RPC utilities (FillPSBT) into
    wallet/psbtwallet.{h,cpp}.
* Switch exceptions from JSONRPCError() to new PSBTException class.
* Split DecodePSBT into DecodeBase64PSBT (old behavior) and DecodeRawPSBT.
* Add one new version of DecodeBase64 utility in strencodings.h (and
    corresponding DecodeBase32 for completeness).
* Factor BroadcastTransaction utility function out of sendrawtransaction RPC
    handler in rpc/rawtransaction.cpp

Note: For those keeping score at home wondering why refactor, this is in anticipation of (and developed in parallel with) a change to actually introduce GUI use of all this stuff, which is already under development and working-ish.

gwillen commented on src/core_read.cpp:
```diff
@@ -176,10 +177,20 @@ bool DecodeHexBlk(CBlock& block, const std::string& strHexBlk)
+
+bool DecodeRawPSBT(PartiallySignedTransaction& psbt, const std::string& tx_data, std::string& error)
+{
+    CDataStream ss_data(tx_data.begin(), tx_data.end(), SER_NETWORK, PROTOCOL_VERSION);
```
Ok, I thought I was very clever converting this from a std::vector<unsigned char> to a std::string, but the reality appears to be that Clang is happy with this line and g++ hates it, so I'm going to have to massage this back towards the original version a bit, in order to get it to build anywhere but macOS.

gwillen commented:
I cleverly made a small change that apparently g++ does not accept but Clang does, so I will have to fix that before it will build on Linux, oops. (I commented on the offending line in the diff.)

gwillen commented:
Hmm, something is also legitimately broken in the tests, will investigate tomorrow. :-\

practicalswift commented on src/psbt.cpp:
```diff
@@ -0,0 +1,275 @@
+bool PartiallySignedTransaction::IsSane() const
+{
+    for (PSBTInput input : inputs) {
+        if (!input.IsSane()) return false;
```
Nit: `std::any_of`? :-)

practicalswift commented on src/psbt.h:
```diff
@@ -0,0 +1,562 @@
+    inline void Unserialize(Stream& s) {
+        // Read loop
+        bool found_sep = false;
+        while(!s.empty()) {
```
Space before `(`. Applies throughout this PR.

Please run all new files through `clang-format` :-)

gwillen replied:
This is moved code, none of this code is new -- I expect I should not actually clang-format it, to preserve move-detection?

promag replied:
Agree.

practicalswift replied:
Makes sense!

practicalswift commented on src/psbt.h:
```diff
@@ -0,0 +1,562 @@
+
+    template <typename Stream>
+    inline void Serialize(Stream& s) const {
+
```
Nit: Drop blank line.

promag commented:
Concept ACK.

> for use in GUI code

Can you detail the GUI part?

gwillen commented:
> Can you detail the GUI part?

I'm adding an interface for offline and multisig signing in the GUI, using PSBT as the on-disk format for the intermediate files. See e.g. how Bitcoin Armory handles offline transactions, which I'm basically copying.

> Is it available?

I pasted it into #bitcoin-core-dev awhile ago -- I will link it here in the comment thread once I'm sure the tip of my branch is still actually building.

jb55 commented:
@gwillen writes:
> * Move most PSBT definitions into psbt.h.
> * Move most PSBT RPC utilities into psbt.{h,cpp}.
> * Move wallet-touching PSBT RPC utilities (FillPSBT) into
>     wallet/psbtwallet.{h,cpp}.
> * Switch exceptions from JSONRPCError() to new PSBTException class.
> * Split DecodePSBT into DecodeBase64PSBT (old behavior) and DecodeRawPSBT.
> * Add one new version of DecodeBase64 utility in strencodings.h (and
>     corresponding DecodeBase32 for completeness).
> * Factor BroadcastTransaction utility function out of sendrawtransaction RPC
>     handler in rpc/rawtransaction.cpp

Is it possible to split this commit into one for each of the points
here? It makes review harder when everything is squashed into a single commit.

gwillen commented:
The GUI work is WIP here: https://github.com/gwillen/bitcoin/tree/feature-offline-v1 . I made that branch at a known-working version, since I'm making significant changes to it right now.

It's not currently based on this PR, which I will have to rebase it onto later; this PR is an extraction of some refactoring that also appears in that PR, plus some other stuff that doesn't yet.

gwillen commented:
@jb55 Yeah, I can definitely split it out if people prefer. Will take a day or two to get to.

Empact commented:
Would be helpful to split the commit up, for review now and for intelligibility when looking back in the future.

achow101 commented:
Concept ACK. Agree with above, split this into multiple commits.

promag commented:
I definitely agree in splitting it.

@gwillen checked out the branch and looks promising.

gwillen committed:
Factor BroadcastTransaction out of sendrawtransaction

Factor out a new BroadcastTransaction function, performing the core work of the
sendrawtransaction rpc, so that it can be used from the GUI code. Move it from
src/rpc/ to src/node/.

gwillen committed:
Move PSBT definitions and code to separate files

Move non-wallet PSBT code to src/psbt.{h,cpp}, and PSBT wallet code to
src/wallet/psbtwallet.{h,cpp}. This commit contains only code movement (and
adjustments to includes and Makefile.am.)

gwillen committed:
Factor out combine / finalize / extract PSBT helpers

Refactor the new CombinePSBT, FinalizePSBT, and FinalizeAndExtractPSBT
general-purpose functions out of the combinepsbt and finalizepsbt RPCs,
for use in the GUI code.

gwillen commented:
Sorry for the delay. I have split the PR into more reasonable commits. The one that does the bulk of the code movement between files ("Move PSBT definitions and code to separate files") is now _only_ movement (and adjusting headers / Makefile), and no other substantive changes. 

The final result is exactly the same as before, except for a single stray blank line I had accidentally introduced in the previous version, which is removed.

Every intermediate step is intended to be sane and self-contained, but I haven't actually verified that every intermediate step cleanly builds yet.

jb55 commented:
This looks much better now! Thanks! 

utACK dd45d3febacb40cf68f58eff08f76db2854fc340

The refactorings look pretty good in general.

I did a quick skim of the code, I didn't see anything obviously wrong but I'm no expert on this part of the codebase.

achow101 commented on src/util/strencodings.cpp:
```diff
@@ -188,9 +188,9 @@ std::vector<unsigned char> DecodeBase64(const char* p, bool* pfInvalid)
 }
 
-std::string DecodeBase64(const std::string& str)
+std::string DecodeBase64(const std::string& str, bool* pfInvalid)
```
nit: variable name should be snake_case. See doc/developer-notes.md for naming convention.

gwillen replied:
This was copied from surrounding context (it just wraps the c_str version of the same function, which has the same parameter named the same thing, and I'm propagating it outwards to the wrapper.)

gwillen replied:
I will fix the argument name, but otherwise leave things alone unless you would strongly prefer the more invasive changes.

Sjors replied:
Yeah, let's not do too much in this PR.

achow101 commented on src/psbt.h:
```diff
@@ -36,6 +36,14 @@ static constexpr uint8_t PSBT_OUT_BIP32_DERIVATION = 0x02;
+/** An exception in PSBT processing that represents a problem with the input data --
+    when the input data is user-supplied, it is appropriate to catch() this and use
+    e.what() to retrieve a human-readable error message. */
+class PSBTException: public std::runtime_error {
```
This results in RPC errors with a code `RPC_MISC_ERROR`. I think it would be better if you could somehow make the error code more specific (the message still comes through).

gwillen replied:
Yeah, I'd love any thoughts on how to accomplish this without gunking up non-RPC code too much when it calls this codepath. I don't really want this to depend on the RPC stuff.

achow101 reviewed (commented):
utACK dd45d3febacb40cf68f58eff08f76db2854fc340

Some comments, nothing that block this IMO

gwillen commented:
FWIW I believe the travis failure is spurious -- it failed a single unrelated test, on a single platform, and the code hasn't changed since the last time it succeeded.

gwillen commented:
(Minor edit to remove a single unnecessary #include line that was bugging me, no significant change.)

promag commented on src/psbt.cpp:
```diff
@@ -232,3 +232,44 @@ bool SignPSBTInput(const SigningProvider& provider, PartiallySignedTransaction&
+    }
+}
+
+PartiallySignedTransaction CombinePSBTs(std::vector<PartiallySignedTransaction> psbtxs) {
```
Commit c062d74

Make the argument a const reference.

nit, brace on new line.

gwillen replied:
Done.

promag reviewed (commented):
utACK c062d74 .Verified the move only commit, just adds necessary includes and updates the makefile with the new file. Some nits and other comments for your consideration.

sipa reviewed (commented):
Big concept ACK

Sjors commented:
@sipa wrote:
> I'm slightly worried about the use of exceptions in code that use usable in more places than just RPC, as C++ can't guarantee that all possible exceptions are being dealt with at the right layer.
> Would an enum with PSBT-related error codes, plus a enum to string conversion function (like in script/script_error) make sense?

I'm also introducing an exception system for ExternalSigner in #14912:

```sh
class ExternalSignerException : public std::runtime_error {
 public:
     using std::runtime_error::runtime_error;
 };

throw ExternalSignerException(strprintf("Some error with detail '%s'", details));
```

I then map that to an RPC error like so:
```sh
    try {
        ExternalSigner::DoSomething(...);
    } catch (const ExternalSignerException& e) {
        throw JSONRPCError(RPC_WALLET_ERROR, e.what());
    }
```

 It's probably a good idea to agree on a single good way to do this.

Sjors commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1030,19 +1030,24 @@ static UniValue sendrawtransaction(const JSONRPCRequest& request)
             + HelpExampleRpc("sendrawtransaction", "\"signedhex\"")
         );
 
-    std::promise<void> promise;
```
Currently in 745c540d05f4a21810eccc5e9450e9b321560560, wrong commit?

gwillen replied:
I think it's the right commit? It moves into BroadcastTransaction.

Sjors commented on src/rpc/rawtransaction.h:
```diff
@@ -13,6 +13,9 @@ namespace interfaces {
 } // namespace interfaces
 
+/** Broadcast a transaction */
+std::string BroadcastTransaction(CTransactionRef tx, bool allowhighfees = false);
```
I also need access to this function in #14912 from other wallet RPC methods. Not really an issue, because I can always include `src/rpc/rawtransaction.h`. But perhaps create a new `src/wallet/transaction.cpp`?

I might add `BroadcastTransaction(PartiallySignedTransaction psbt, bool allowhighfees = false)` in my own PR.

gwillen replied:
Agreed that this should be elsewhere, I will create src/wallet/transaction.cpp for it. (I was reluctant to be too original just to move a single function, but I prefer it that way also.)

Sjors replied:
(agree with the latter, I ended up with a different approach too, the seperate file is useful enough)

Sjors reviewed (commented):
Reviewed c062d74e1187e481960ca6868a46642b8bbfbd2e, mostly happy.

Sjors reviewed (commented):
Lightly tested c062d74 using #14912, seems to work well.

Needs trivial rebase due because #14906 changed `PartiallySignedTransaction::PartiallySignedTransaction(const CTransaction& tx) : tx(tx)` to `CMutableTransaction` in `src/script/sign.{h/cpp}`

gwillen commented:
@promag @sipa @Sjors Can you have an argument, here or elsewhere, about how error handling should work, and then tell me what to do? :-) I am ambivalent between "better exceptions" and "switch to return values".

gwillen committed:
Add pf_invalid arg to std::string DecodeBase{32,64}

Add support for the optional "pf_invalid" out parameter (which allows the caller
to detect decoding failures) to the std::string versions of DecodeBase32 and
DecodeBase64. The char* versions already have this feature.

Also, rename all uses of pfInvalid to pf_invalid to match style guidelines.

gwillen commented on src/core_read.cpp:
```diff
@@ -176,10 +176,20 @@ bool DecodeHexBlk(CBlock& block, const std::string& strHexBlk)
+
+bool DecodeRawPSBT(PartiallySignedTransaction& psbt, const std::string& tx_data, std::string& error)
+{
+    CDataStream ss_data(tx_data.data(), tx_data.data() + tx_data.size(), SER_NETWORK, PROTOCOL_VERSION);
```
Is this safe (does it not "prove too much")? Certainly I can imagine it might cause worse error messages in some cases. I assume it wasn't done that way for some reason originally?

Sjors replied:
@achow101 thoughts?

gwillen replied:
Just for fun, I went and checked -- this code is essentially unchanged since the repo was first imported to github in 2009. So the rationale is probably lost to history...

gwillen commented on src/core_io.h:
```diff
@@ -37,7 +37,8 @@ bool DecodeHexBlockHeader(CBlockHeader&, const std::string& hex_header);
 bool ParseHashStr(const std::string& strHex, uint256& result);
 std::vector<unsigned char> ParseHexUV(const UniValue& v, const std::string& strName);
-NODISCARD bool DecodePSBT(PartiallySignedTransaction& psbt, const std::string& base64_tx, std::string& error);
+NODISCARD bool DecodeBase64PSBT(PartiallySignedTransaction& psbt, const std::string& base64_tx, std::string& error);
```
Done.

gwillen commented on src/core_io.h:
```diff
@@ -37,7 +37,8 @@ bool DecodeHexBlockHeader(CBlockHeader&, const std::string& hex_header);
 std::vector<unsigned char> ParseHexUV(const UniValue& v, const std::string& strName);
-NODISCARD bool DecodePSBT(PartiallySignedTransaction& psbt, const std::string& base64_tx, std::string& error);
+NODISCARD bool DecodeBase64PSBT(PartiallySignedTransaction& psbt, const std::string& base64_tx, std::string& error);
+NODISCARD bool DecodeRawPSBT(PartiallySignedTransaction& psbt, const std::string& tx_data, std::string& error);
```
Done.

gwillen commented on src/core_read.cpp:
```diff
@@ -176,10 +176,20 @@ bool DecodeHexBlk(CBlock& block, const std::string& strHexBlk)
+    bool invalid;
+    std::string tx_data = DecodeBase64(base64_tx, &invalid);
+    if (invalid) {
+        error = "invalid base64";
```
Done.

promag commented:
@gwillen I'd avoid using exceptions to control flow. For instance, you have:
```cpp
bool FillPSBT(const CWallet* pwallet, PartiallySignedTransaction& psbtx, int sighash_type = 1 /* SIGHASH_ALL */, bool sign = true, bool bip32derivs = false)
```
And looks like it returns `false` if it fails, but that's wrong. In fact, it throws if some error occurs and returns whether the transaction has a complete set of signatures.

Looking the existing calls you see the semantic difference:
https://github.com/bitcoin/bitcoin/blob/7275365c9bc7e7ebd6bbf7dcb251946aac44b5de/src/wallet/rpcwallet.cpp#L4035
https://github.com/bitcoin/bitcoin/blob/7275365c9bc7e7ebd6bbf7dcb251946aac44b5de/src/wallet/rpcwallet.cpp#L4149

Empact commented on src/core_io.h:
```diff
@@ -37,7 +37,11 @@ bool DecodeHexBlockHeader(CBlockHeader&, const std::string& hex_header);
+
+// Decode a base64ed PSBT into a PartiallySignedTransaction
+NODISCARD bool DecodeBase64PSBT(PartiallySignedTransaction& decoded_psbt, const std::string& base64_psbt, std::string& error);
+// Decode a raw (binary blob) PSBT into a PartiallySignedTransaction
```
nit: use `//!` or equivalent for doxygen

gwillen replied:
Aha, right. Done.

gwillen committed:
Split DecodePSBT into Base64 and Raw versions

Split up DecodePSBT, which both decodes base64 and then deserializes a
PartiallySignedTransaction, into two functions: DecodeBase64PSBT, which retains
the old behavior, and DecodeRawPSBT, which only performs the deserialization.

Add a test for base64 decoding failure.

gwillen commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1030,19 +1030,24 @@ static UniValue sendrawtransaction(const JSONRPCRequest& request)
         throw JSONRPCError(RPC_DESERIALIZATION_ERROR, "TX decode failed");
     CTransactionRef tx(MakeTransactionRef(std::move(mtx)));
+
+    bool allowhighfees = !request.params[1].isNull() && request.params[1].get_bool();
```
Makes sense, done.

gwillen commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1030,19 +1030,24 @@ static UniValue sendrawtransaction(const JSONRPCRequest& request)
+    return BroadcastTransaction(tx, allowhighfees);
+}
+
+std::string BroadcastTransaction(CTransactionRef tx, bool allowhighfees) {
```
Can you clarify? These are both by-value, so I wouldn't ordinarily bother marking them const. Am I supposed to?

sipa replied:
There is little point in doing so, as constness of by-value arguments doesn't even affect the function signature.

You can if you really want to make sure the local variables associated with them don't change during the function execution, but that's unusual in our codebase, and not very important.

sipa commented:
@gwillen My preference is an enum over exceptions.

gwillen commented on src/psbt.h:
```diff
@@ -391,20 +391,15 @@ struct PartiallySignedTransaction
-    {
-        return !(a == b);
+    // Checks if they refer to the same underlying CTransaction
+    bool SameTx(const PartiallySignedTransaction& psbt) {
```
Makes sense, done.

gwillen commented on src/psbt.h:
```diff
@@ -391,20 +391,15 @@ struct PartiallySignedTransaction
 
     bool IsNull() const;
-    void Merge(const PartiallySignedTransaction& psbt);
+    NODISCARD bool Merge(const PartiallySignedTransaction& psbt);
```
Done.

gwillen committed:
Remove op== on PSBTs; check compatibility in Merge

Remove the op== on PartiallySignedTransaction, which only checks that the
CTransactions are equal. Instead, check this directly in Merge, and return
false if the CTransactions are not equal (so the PSBTs cannot be merged.)

gwillen commented on src/psbt.h:
```diff
@@ -0,0 +1,562 @@
+/** Signs a PSBTInput, verifying that all provided data matches what is being signed. */
+bool SignPSBTInput(const SigningProvider& provider, PartiallySignedTransaction& psbt, int index, int sighash = SIGHASH_ALL);
+
+/** Finalizes a PSBT if possible, combining partial signatures and optionally extracting to a signed transaction ready for sending. */
```
>Though maybe that's too much refactoring for this PR.

No, it's too late, you've got me thinking about it now. The "finalize and maybe extract" interface is kind of weird as it stands, and the string-handling stuff belongs in the rpc handler, not in the general-purpose functions. I clearly drew the abstraction boundaries in the wrong place.

I've reworked this in a way that feels a bit better to me. It's not perfect but it no longer obviously feels to me like I can improve it.

gwillen commented:
I think I've addressed everything except the error-handling / exceptions stuff, which I'm still thinking about and will punt to tomorrow.

I significantly reworked the FinalizePSBT interface, so it deserves another look. Everything else was relatively minor.

Sjors reviewed (commented):
utACK f124a51 modulo the broken bench build on Travis (can't reproduce that on macOS)

Enum error handling works for me too.

gwillen commented:
Ok, I ended up doing something that feels a little gross and unorthodox for the error handling.

The problems with a new enum are several-fold: Most notably, it would be at least two new enums, since BroadcastTransaction and the various PSBT functions are not really of the same sort, unless we want something like a TransactionError type (which I wouldn't be against.) Aside from that, the ONLY consumer of error codes in any form (as far as I can tell) would be RPC callers, which means the only point of making a bunch of new error constants is to immediately turn around and define a mapping of them to RPC error constants.

(ScriptErrors get handled a little differently, as far as I can tell -- they just get converted to strings when they are stuffed into RPC responses. As far as I can tell, the _only_ other use for them is in tests. If we were going to go that route here, I would just advocate for pure string errors, since the constants wouldn't have any interesting producers or consumers at that point.)

Anyway, on the theory that the only consumer of error codes is the RPC subsystem, my approach is this: I retained the use of RPCErrorCode and JSONRPCError. Instead of throwing the resulting UniValue objects, I am using out parameters for them. Then when I am back in the RPC codepath, I take the object and throw it. In the GUI codepath, I expect to extract the error message from it to display to the user, and ignore the code.

This is definitely a little bit crazy, but I was having trouble coming up with any better ideas. Thoughts?

gwillen commented:
Oh, and I think the Travis failure is real -- I think the problem is that actually BroadcastTransaction does NOT belong under wallet/ and needs to be linked even when the wallet is disabled. So right now disabling the wallet causes the link to fail.

Suggestions for where to move it? Right now it's a pair of new files named wallet/transaction.{h,cpp}.

Sjors commented:
> The ONLY consumer of error codes in any form (as far as I can tell) would be RPC callers,

In #14912 I plan to move more code into the `ExternalSigner` class, which will be used by both RPC and the GUI.

> general-purpose TransactionError type

Works for me

> that actually BroadcastTransaction does NOT belong under `wallet/`

Indeed it doesn't. Nodes without a wallet can also broadcast transactions and do various things with PSBT. Maybe `src/node/transaction.{h,c}`? @ryanofsky what makes sense given #10973? I'm assuming it doesn't really belong in `src/interfaces`?

Sjors commented on src/wallet/psbtwallet.cpp:
```diff
@@ -2,14 +2,14 @@
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
-#include <rpc/protocol.h>
+#include <rpc/protocol.h> // For JSONRPCError and RPCErrorCode
```
Yuck circular dependency? :-)

ryanofsky commented:
> > that actually BroadcastTransaction does NOT belong under wallet/

> Indeed it doesn't. Nodes without a wallet can also broadcast transactions and do various things with PSBT. Maybe src/node/transaction.{h,c}? @ryanofsky what makes sense given #10973? I'm assuming it doesn't really belong in src/interfaces?

It would be reasonable to make `BroadcastTransaction` an `interfaces::Chain` method. (`Chain` is the object a wallet process uses to communicate with a node process in #10102), and I've actually kind of done this already in 9eee26ef6071451ee426f71beb6b77184bc8b073 from #15288. It would also be reasonable to move `BroadcastTransaction` somewhere else if it can't stay in wallet because it will be used by non-wallet code. Whatever you decide should be fine.

gwillen commented:
@ryanofsky The reason I need something like BroadcastTransaction is that, currently, none of the interfaces:: stuff exports a way to broadcast a transaction that didn't come from the wallet (and I need the RPC-like ability to do it with a blob I was handed.) So probably these codepaths should eventually merge! :-)

Sjors commented:
By the way `wallet/transaction.{h,cpp}` is still useful, even if not for this PR. E.g. I have a need to move `FundTransaction(CWallet* const pwallet, CMutableTransaction& tx...` out of `rpcwallet.{h,cpp}`

I would probably also wrap that inside a `FundPSBT` method in `wallet/psbtwallet.{h,cpp}`, so that `walletcreatefundedpsbt` doesn't have to deal directly with `CMutableTransaction`.

gwillen commented:
@ryanofsky Why interfaces::Chain and not interfaces::Node? That seems like a better fit. (Especially since the GUI ClientModel object has a Node interface handy, but I don't see where the GUI easily gets access to a Chain, although I'm not totally sure where to look.)

gwillen commented:
Oh, good point. So I'm going ahead with creating src/node for it to live in, and then later I will forward from interfaces::Node (when I actually add a GUI use of it.)

gwillen commented:
Ok, I think I have fixed the issue with BroadcastTransaction linking, by moving it out of the wallet stuff.

I have made another attempt at the error stuff. :-) This time we have a TransactionError enum, and a translation table from TransactionError to RPCErrorCode, as well as strings for each TransactionError.

The result is _almost_ faithful, for RPC clients, to how it was before, except:
- I have reclassified syntactically-valid but semantically-invalid PSBTs, from RPC_DESERIALIZATION_ERROR to RPC_INVALID_PARAMETER, in one case (we were being inconsistent)
- When AcceptToMemoryPool fails, we can extract a string from the failure using FormatStateMessage. But we no longer have an easy way to pass that message back to the RPC caller, so right now we don't. Probably I should add another out param to fix this or something.

gwillen commented:
In fact, it turns out that changing the error message wording of BroadcastTransaction at all breaks a bunch of functional tests. Working on it.

gwillen commented:
There we go! Ready for another look.

sipa commented on src/core_io.h:
```diff
@@ -37,7 +37,11 @@ bool DecodeHexBlockHeader(CBlockHeader&, const std::string& hex_header);
+//! Decode a base64ed PSBT into a PartiallySignedTransaction
+NODISCARD bool DecodeBase64PSBT(PartiallySignedTransaction& decoded_psbt, const std::string& base64_psbt, std::string& error);
+//! Decode a raw (binary blob) PSBT into a PartiallySignedTransaction
+NODISCARD bool DecodeRawPSBT(PartiallySignedTransaction& decoded_psbt, const std::string& raw_psbt, std::string& error);
```
It seems a bit strange that `raw_psbt` is a string and not a `std::vector<unsigned char>`. The `DecodeBase64` call below also returns a `std::vector<unsigned char>`, so changing this would avoid a conversion (which I'm a bit surprised happens automatically).

gwillen replied:
In fact, DecodeBase64 does not return a vector, it does return a std::string when given a std::string, which I imagine is why I ended up doing it this way, to avoid having to convert it? (When given a char*, it returns a vector. Your guess why it was done this way is as good as mine.)

I did carefully check that std::string is just a binary-safe as vector<char>. I'm fine leaving it or changing it as you prefer.

sipa replied:
Oh, you're right. That's weird, but probably out of scope for this PR. I'm fine with leaving this as is.

sipa commented on src/rpc/util.cpp:
```diff
@@ -139,6 +139,8 @@ RPCErrorCode RPCErrorFromTransactionError(TransactionError terr)
         case TRANSACTION_ERR_P2P_DISABLED:
             return RPC_CLIENT_P2P_DISABLED;
         case TRANSACTION_ERR_INVALID_PSBT:
+        case TRANSACTION_ERR_PSBT_MISMATCH:
```
This changes the error code for TRANSACTION_ERR_INVALID_PSBT from RPC_DESERIALIZATION_ERROR to RPC_INVALID_PARAMETER. Is that intentional?

gwillen replied:
Yes, because there are two cases where we return an RPC error for a PSBT being semantically invalid, and we were returning DESERIALIZATION_ERROR in one case (when the invalid PSBT was given as a parameter) and INVALID_PARAMETER in the other (when the invalid PSBT was the result of a merge), and the I though the latter seemed more reasonable. My sense was that DESERIALIZATION_ERROR is only meant for syntactically invalid arguments? I'd be willing to change it back though. It would be easier if the two cases were treated the same.

sipa replied:
I don't have any strong opinion on the RPC coded. I just wanted to make sure this was intentional (as it could easily have been a copypaste error or so).

sipa reviewed (commented):
Concept ACK. Also rough code review ACK, though I haven't verified code movement yet.

sipa commented on src/node/transaction.h:
```diff
@@ -0,0 +1,43 @@
+#include <primitives/transaction.h>
+#include <uint256.h>
+
+typedef enum TransactionError_t {
```
This looks like a C-ism. You can just use `enum TransactionError { ... };` in C++ with (I think) the same effect.

gwillen replied:
In my defense I copied it from ScriptError_t. I'll change it.

sipa commented on src/psbt.h:
```diff
@@ -0,0 +1,574 @@
+bool FinalizePSBT(PartiallySignedTransaction& psbtx);
+
+/**
+ * Finalizes a PSBT if possible, and extracts it to a CTransaction if it could be finalized.
```
Nit: `CMutableTransaction`.

gwillen replied:
Oh, you're right, I think I had to change it for #14906 and forgot to change the comment. Will fix.

sipa reviewed (commented):
utACK acac678ebe6e105e0a775b6eff37b1583139bf21, only nits.

practicalswift commented on src/node/transaction.h:
```diff
@@ -0,0 +1,43 @@
+#include <primitives/transaction.h>
+#include <uint256.h>
+
+enum TransactionError {
```
Should be `enum class`?

gwillen replied:
Oh, I've never used that before but it looks cool.

It looks like if I made that change, all references to the enum constants would have to be qualified with the name of the enum? So I should remove all the TRANSACTION_ERR prefixes, so e.g. TRANSACTION_ERR_MISSING_INPUTS would become TransactionError::MISSING_INPUTS?

I'm always in favor of adding more compile-time checks when it's easy, so I'm happy to make the change if that's the preferred approach.

gwillen replied:
Updated.

gwillen committed:
Switch away from exceptions in refactored tx code

After refactoring general-purpose PSBT and transaction code out of RPC code,
for use in the GUI, it's no longer appropriate to throw exceptions. Instead we
now return bools for success, and take an output parameter for an error object.
We still use JSONRPCError() for the error objects, since only RPC callers
actually care about the error codes.

Sjors commented:
re-utACK 6097ae0, main changes since my last review:
* moved `BroadcastTransaction` to `src/node` instead of `src/wallet/transaction.cpp`, because it's not a wallet thing.
* switched away from exceptions d988f06
* `CombinePSBTs` now returns boolean and has an `&out` and `&error` param

gwillen commented:
Rebased.

gwillen commented:
Can 1-2 people who already acked, but haven't looked in awhile, take a quick look and give a reack? I do not expect further changes unless I am forced to rebase again.

achow101 commented:
utACK 102faad81efa1cb12c29c466cfe81fc8c7351e1d

MeshCollider commented:
@gwillen thanks for being so proactive with addressing review comments :)
